tools/flight_recorder/components/utils.py

```diff
@@ -0,0 +1,230 @@
+
+
+def find_coalesced_group(
+    pg_name: str, entries: List[Dict[str, Any]]
+) -> List[Tuple[int, Dict[str, Any]]]:
+    """Given a list of entries, if the collective_seq_id of the first entry matches that of subsequent ones,
+    build an return a list of entries terminating in a 'coalesced' op entry all sharing a collective_seq_id
+    TODO: handle p2p_seq_id v/s collective_seq_id separately here.
+    """
+    found = []
+    collective_seq_id = None
+    for i, e in enumerate(entries):
+        if e["process_group"][0] != pg_name:
+            continue
+        elif collective_seq_id is None:
+            collective_seq_id = e["collective_seq_id"]
+            found.append((i, e))
+        elif e["collective_seq_id"] == collective_seq_id:
+            found.append((i, e))
+        else:
+            break
+
+    if len(found) > 1:
+        assert found[-1][1]["profiling_name"] == "nccl:coalesced"
+        return found
+    return []
+
+
+def just_print_entries(
+    all_entries: Dict[int, List[Dict[str, Any]]],
+    _groups: Dict[str, Group],
+    _memberships: Dict[str, Set[Any]],
+) -> None:
+    rows = []
+    ranks = sorted(all_entries.keys())
+    headers = [f"Rank {rank}" for rank in ranks]
+    progress = True
+    while progress:
+        progress = False
+        row = []
+        for rank in ranks:
+            if len(all_entries[rank]) == 0:
+                row.append("")
+            else:
+                entry = all_entries[rank].pop(0)
+                row.append(str(Op(entry, _memberships)))
+                progress = True
+        if progress:
+            rows.append(row)
+
+    print(tabulate(rows, headers=headers))  # type: ignore[operator]
+
+
+def check_no_missing_dump_files(
+    entries: Dict[str, Any], memberships: List[Membership]
+) -> None:
+    all_ranks = set()
+    for membership in memberships:
+        all_ranks.add(str(membership.global_rank))
+    dumps_ranks = set(entries.keys())
+    assert (
+        dumps_ranks == all_ranks
+    ), f"Missing dump files from ranks {all_ranks - dumps_ranks}"
+
+
+def check_version(versions: Dict[str, Any]) -> None:
+    for rank, version in versions.items():  # noqa: PERF102
+        major, minor = map(int, version.split("."))
+        # assert major == 2, f"Rank {rank} unsupported version {version}"
+        # assert minor >= 0, f"Rank {rank} unsupported version {version}"
+
+
+def check_trace_from_beginning(entries: Dict[str, Any]) -> bool:
+    for rank in entries:
+        first_record_id = entries[rank][0]["record_id"]
+        # TODO add more sequence information such that analysis can proceed even without complete buffer
+
+        # assert first_record_id == 0, f"Rank {rank} trace does not start at time 0 (first record is {first_record_id}."
+        if first_record_id != 0:
+            print(
+                f"Rank {rank} trace does not start at time 0 (first record is {first_record_id}."
+            )
+            return False
```

Comment:
I'm not sure if this is correct - but we can fix it later.
Since it is a circular buffer, more often than not, the first record id will not be `0`.
The important thing is that flight recorder records are ordered. That is, 0th record will have smallest record id and Nth record (where N is buffer size) will have largest record id.

Reply:
right, that makes sense.